### PR TITLE
CSRF Implementation

### PR DESCRIPTION
**Pull request by therealssj**
1. Add CSRF check to all POST request
2. Add `/csrf` endpoint to fetch the current csrf token or refresh it if expired
3. Add `-disable-csrf` flag 
4. Update existing tests

**Tasks**
- [x] Backend
- [x] Frontend
- [ ] Tests

Resolves: #896

### CODE REVIEW

File: src/daemon/gateway.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 type GatewayConfig struct {
 	BufferSize       int
 	DisableWalletAPI bool
+	DisableCSRF      bool
```
Comment (gz-c) [inline src/daemon/gateway.go]: csrf has no relationship to the daemon gateway

File: src/gui/README.md [hunk 2]
```diff
@@ -38,6 +38,25 @@ result:
...
+example:
+
+```sh
+curl http://127.0.0.1:6420/csrf
+```
+
+result:
+
+```
+EehRw-67K-HUzzI9ph3-a06O3bgx-4ub9SEunPGekW-5J33YpH8znybQj5zhT0z7o8us4K_q8LTSDNLTcZ4QbQ
```
Comment (gz-c) [inline src/gui/README.md]: Should return `{"csrf_token":"..."}` with `Content-Type` "application_json"

File: src/gui/csrf.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	"time"
+
+	"github.com/skycoin/skycoin/src/util/utc"
+
+	wh "github.com/skycoin/skycoin/src/util/http" //http,json helpers
+)
+
+const (
+	// the name of CSRF header
+	HeaderName = "X-CSRF-Token"
```
Comment (gz-c) [inline src/gui/csrf.go]: `CSRFHeaderName`

File: src/gui/csrf.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Max-Age in seconds for cookie. 30 seconds default
+	MaxAge = time.Duration(30) * time.Second
```
Comment (gz-c) [inline src/gui/csrf.go]: `CSRFMaxAge`

File: src/gui/csrf.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+	// csrf token length
+	tokenLength = 64
```
Comment (gz-c) [inline src/gui/csrf.go]: `csrfTokenLength`

File: src/gui/csrf.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+				return
+			}
+		}
+
+		handler.ServeHTTP(w, r)
+	})
+}
+
+// b64encode returns a url safe base64 encoded string
+func b64encode(data []byte) string {
```
Comment (gz-c) [inline src/gui/csrf.go]: superfluous

File: src/gui/csrf.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	token := CSRFToken{
+		bytes,
+		utc.Now().Add(MaxAge),
+	}
+
+	return &token
+}
+
+func checkForPRNG() {
```
Comment (gz-c) [inline src/gui/csrf.go]: not needed

File: src/gui/csrf.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+// b64encode returns a url safe base64 encoded string
+func b64encode(data []byte) string {
+	return base64.RawURLEncoding.EncodeToString(data)
+}
+
+// generateToken generates a new CSRF Token
+func generateToken() *CSRFToken {
+	bytes := make([]byte, tokenLength)
+
+	if _, err := io.ReadFull(rand.Reader, bytes); err != nil {
```
Comment (gz-c) [inline src/gui/csrf.go]: use `cipher.RandByte`

File: src/gui/csrf.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+type CSRFStore struct {
+	token   *CSRFToken
+	enabled bool
+	sync.Mutex
+	once sync.Once
+}
+
+// getTokenValue returns a url safe base64 encoded token
+func (c *CSRFStore) getTokenValue() string {
+	return b64encode(c.token.Value)
```
Comment (gz-c) [inline src/gui/csrf.go]: Need to lock the mutex, can use `sync.RWMutex`

File: src/gui/csrf.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+// so there is no explicit method to just set the value of the token
+func (c *CSRFStore) setToken(token *CSRFToken) {
+	c.Lock()
+	defer c.Unlock()
+	c.token = token
+}
+
+// verifyExpireTime checks if token expiry time is greater than current time
+func (c *CSRFStore) verifyExpireTime() bool {
+	return utc.UnixNow() < c.token.ExpiresAt.Unix()
```
Comment (gz-c) [inline src/gui/csrf.go]: Use `time.After` for comparison

UTC time doesn't matter here either

File: src/gui/csrf.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+}
+
+// verifyToken checks that the given token is same as the internal token
+func (c *CSRFStore) verifyToken(headerToken string) bool {
+	// check if token is initialized
```
Comment (gz-c) [inline src/gui/csrf.go]: needs to use locks

File: src/gui/csrf.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+		// make sure token is still valid
+		return c.verifyExpireTime()
+	}
+
+	return false
+}
+
+// getCSRFStore returns a CSRFStore instance
+// initializes the csrf store if not already done
+func (c *CSRFStore) getCSRFStore() *CSRFStore {
```
Comment (gz-c) [inline src/gui/csrf.go]: unnecessary

File: src/gui/http.go [hunk 13]
```diff
@@ -132,6 +132,13 @@ func NewServerMux(host, appLoc string, gateway Gatewayer) *http.ServeMux {
 		webHandler(route, http.FileServer(http.Dir(appLoc)))
 	}
 
+	var CSRF = new(CSRFStore)
```
Comment (gz-c) [inline src/gui/http.go]:
```go
csrfStore := &CSRFStore{
    Enabled: !disableCSRF,
}
```

File: src/gui/csrf.go [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+	})
+
+	return c
+}
+
+// method: GET
+// url: /csrf
+func getCSRFToken(gateway Gatewayer, store *CSRFStore) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet || !store.enabled {
```
Comment (gz-c) [inline src/gui/csrf.go]: Return a 404 if the store is disabled and document that in the readme and function docstring

File: src/gui/csrf.go [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+			wh.Error405(w)
+			return
+		}
+
+		// check if token is still valid
```
Comment (gz-c) [inline src/gui/csrf.go]: always generate a new token

File: src/gui/http.go [hunk 16]
```diff
@@ -188,12 +195,13 @@ func NewServerMux(host, appLoc string, gateway Gatewayer) *http.ServeMux {
 	webHandler("/wallet/update", walletUpdateHandler(gateway))
```
Comment (gz-c) [inline src/gui/http.go]: wrap all handlers with `CSRFCheck` except for the csrf handler

File: src/gui/csrf.go [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+
+		wh.SendOr404(w, store.getTokenValue())
+
+	}
+}
+
+// CSRFCheck verifies X-CSRF-Token header value
+func CSRFCheck(handler http.Handler, store *CSRFStore) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if store.enabled {
```
Comment (gz-c) [inline src/gui/csrf.go]: only check csrf if the `r.Method == http.MethodPost`

File: src/gui/csrf.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+	return base64.RawURLEncoding.EncodeToString(c.token.Value)
+}
+
+// setToken sets a new CSRF token
+// if the value is changing the expire time should also change
+// so there is no explicit method to just set the value of the token
+func (c *CSRFStore) setToken(token *CSRFToken) {
+	c.Lock()
+	defer c.Unlock()
+	c.token = token
```
Comment (iketheadore) [inline src/gui/csrf.go]: It's not safe to assign pointer here, the `token` argument can be modified outside of the function.

**Comment by samuelvisscher** [comment]
Also fixes: host problems with ng serve

File: src/gui/http.go [hunk 19]
```diff
@@ -174,20 +178,21 @@ func NewServerMux(host, appLoc string, gateway Gatewayer) *http.ServeMux {
 	//  fee: Number of hours to use as fee, on top of the default fee.
 	//  Returns total amount spent if successful, otherwise error describing
 	//  failure status.
-	webHandler("/wallet/spend", walletSpendHandler(gateway))
+	webHandler("/wallet/spend", CSRFCheck(walletSpendHandler(gateway), csrfStore))
```
[the remark this answers is not in the document]
Comment (therealssj) [reply src/gui/http.go]: damn misunderstood what you meant last time

File: src/gui/csrf.go [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	return false
+}
+
+// method: GET
+// url: /csrf
+func getCSRFToken(gateway Gatewayer, store *CSRFStore) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodGet || !store.Enabled {
```
[the remark this answers is not in the document]
Comment (therealssj) [reply src/gui/csrf.go]: in previous comment you said change it to 404

File: src/gui/http.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (gz-c) [reply src/gui/http.go]: *except the csrf endpoint

**Comment by gz-c** [comment]
#937

**Pull request closed without merging**